packages/lisk-transactions/src/transactions/3_vote_transaction.ts

```diff
@@ -0,0 +1,478 @@
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import BigNum from 'browserify-bignum';
+import { VOTE_FEE } from '../constants';
+import { TransactionError, TransactionMultiError } from '../errors';
+import { Account, Status, TransactionJSON } from '../transaction_types';
+import { prependMinusToPublicKeys, prependPlusToPublicKeys } from '../utils';
+import {
+	isTypedObjectArrayWithKeys,
+	validateAddress,
+	validatePublicKeys,
+	validator,
+} from '../utils/validation';
+import {
+	Attributes,
+	BaseTransaction,
+	createBaseTransaction,
+	CreateBaseTransactionInput,
+	ENTITY_ACCOUNT,
+	EntityMap,
+	RequiredState,
+	TransactionResponse,
+} from './base';
+
+const PREFIX_UPVOTE = '+';
+const PREFIX_UNVOTE = '-';
+const MAX_VOTE_PER_ACCOUNT = 101;
+const MIN_VOTE_PER_TX = 1;
+const MAX_VOTE_PER_TX = 33;
+
+export interface VoteAsset {
+	readonly votes: ReadonlyArray<string>;
+}
+
+export interface CreateVoteAssetInput {
+	readonly unvotes?: ReadonlyArray<string>;
+	readonly votes?: ReadonlyArray<string>;
+}
+
+export type CastVoteInput = CreateBaseTransactionInput & CreateVoteAssetInput;
+
+export interface RequiredVoteState extends RequiredState {
+	readonly dependentState?: {
+		readonly [ENTITY_ACCOUNT]: ReadonlyArray<Account>;
+	};
+}
+
+export const voteAssetTypeSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			items: {
+				type: 'string',
+			},
+		},
+	},
+};
+
+export const voteAssetFormatSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			uniqueSignedPublicKeys: true,
+			minItems: MIN_VOTE_PER_TX,
+			maxItems: MAX_VOTE_PER_TX,
+			items: {
+				type: 'string',
+				format: 'signedPublicKey',
+			},
+		},
+	},
+};
+
+const validateInputs = ({
+	votes = [],
+	unvotes = [],
+}: CreateVoteAssetInput): void => {
+	if (!Array.isArray(votes)) {
+		throw new Error(
+			'Please provide a valid votes value. Expected an array if present.',
+		);
+	}
+	if (!Array.isArray(unvotes)) {
+		throw new Error(
+			'Please provide a valid unvotes value. Expected an array if present.',
+		);
+	}
+	validatePublicKeys([...votes, ...unvotes]);
+};
+
+export class VoteTransaction extends BaseTransaction {
+	public readonly containsUniqueData = true;
+	public readonly asset: VoteAsset;
+	public readonly fee: BigNum = new BigNum(VOTE_FEE);
+
+	public constructor(tx: TransactionJSON) {
+		super(tx);
+		const typeValid = validator.validate(voteAssetTypeSchema, tx.asset);
+		const errors = validator.errors
+			? validator.errors.map(
+					error =>
+						new TransactionError(
+							`'${error.dataPath}' ${error.message}`,
+							tx.id,
+							error.dataPath,
+						),
+			  )
+			: [];
+		if (!typeValid) {
+			throw new TransactionMultiError('Invalid field types', tx.id, errors);
+		}
+		this.asset = tx.asset as VoteAsset;
+	}
+
+	public static create(input: CastVoteInput): object {
```

Comment:
shouldn't this function return TransactionJSON?

Reply:
we can't return TransactionJSON because without passphrase, it's not enough to become a transaction ='(